File: 1544-make-the-string-great/1544-make-the-string-great.java
class Solution {
    public boolean check(Deque<Character> stack, char b){
        char a = stack.peek();
        
        if(a >= 'A' && a <= 'Z' && b >= 'a' && b <= 'z' && Character.toLowerCase(a) == b)    return true;
        
        else if(a >= 'a' && a <= 'z' && b >= 'A' && b <= 'Z' && Character.toUpperCase(a) == b)   return true;
        
        return false;
    }
    public String makeGood(String s) {
        int n = s.length();
        Deque<Character> stack = new ArrayDeque<>();
        
        for(int i = 0; i < n;){
            while(i < n && !stack.isEmpty() && check(stack, s.charAt(i))){
                stack.pop();
                i++;
            }
            
            if(i < n)   stack.push(s.charAt(i++));
        }
        
        StringBuilder res = new StringBuilder();
        
        while(!stack.isEmpty()){
            res.append(stack.peek());
            stack.pop();
        }
        
        res.reverse();
        
        return res.toString();
    }
}

File: 1544-make-the-string-great/1544-make-the-string-great.cpp
class Solution {
public:
    bool check(stack<char>& stk, char b) {
        if (!stk.empty()) {
            char a = stk.top();
            if ((a >= 'A' && a <= 'Z' && b >= 'a' && b <= 'z' && tolower(a) == b) ||
                (a >= 'a' && a <= 'z' && b >= 'A' && b <= 'Z' && toupper(a) == b))
                return true;
        }
        return false;
    }

    string makeGood(string s) {
        int n = s.length();
        stack<char> stk;

        for (int i = 0; i < n;) {
            while (i < n && !stk.empty() && check(stk, s[i])) {
                stk.pop();
                i++;
            }
            if (i < n)
                stk.push(s[i++]);
        }

        string res = "";
        while (!stk.empty()) {
            res += stk.top();
            stk.pop();
        }

        reverse(res.begin(), res.end());

        return res;
    }
};